job/server/src/main/java/alluxio/job/plan/transform/format/JobPath.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.job.plan.transform.format;
+
+import alluxio.client.ReadType;
+import alluxio.client.WriteType;
+import alluxio.conf.InstancedConfiguration;
+import alluxio.conf.PropertyKey;
+
+import com.google.common.base.Objects;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.security.UserGroupInformation;
+
+import java.io.IOException;
+import java.net.URI;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Implementation of {@link Path} that has a cache for getting file system.
+ */
+public class JobPath extends Path {
+
+  private static final ConcurrentHashMap<FileSystemKey, FileSystem> CACHE =
```

Comment:
should we use a guava `CacheBuilder` with a maximum size, to limit the unlimited growth?

Reply:
For now, this is not much of a concern. mScheme, mAuthority, and mUgi are all basically always gonna be constant and even mReadType and mWriteType is very limited.

What's going to get really annoying once we limit size is I don't know which cache is going to need to close the FileSystem on removal.